### ISSUE TITLE
[IDEA] Alternate model syntax

### ISSUE THREAD

**Issue by grigi** [tortoise/tortoise-orm#71]
In trying to solve the DB → python type mapping, and the advent of `typing`, a possible solution is to define types that then get evaluated into both Python and DB types:

This also raises the question of a cleaner (more modern) model definition, by defining the type:

Since py3.6 we have the typed variables:
```py3
class Author(Model):
    name: str[255]
    books: reverse_fk[Book]

class Book(Model):
    title: str[255]
    author: Author
    published: datetime
    logo: bytes = None
    pages: int
    extra: JSONField = {}
    price: Decimal[18,5]
    created_at: datetime[auto_now_add]
```
Which is already standard in Py3.7 [@dataclass](https://docs.python.org/3/library/dataclasses.html)

We could extend type definitions so we can build model definitions off of types.
e.g. `str[100] = 'some default'` → `CharField(max_length=100, default='some default')`
The syntax is not comprehensive, but we could probably use type declarations for a large portion of use cases.

This would require Py3.6+ to function, so dropping Py3.5 & PyPy.

**Comment by abondar** [comment]
Even though possibility is neat, I don't see any real life advantages of implementing this. 

It would be harder for us to implement both syntaxes and development of standard syntax could be held down because of type annotation limitations.

And for users, having two ways to do same thing would be confusing and will most likely lead to wrong understanding of how to use Tortoise.

**Comment by grigi** [comment]
Yes, hence the "wild idea".

To get IDE's to understand how models are configured we would need to mess with `typing` in any case.
Here we define python types that get mapped to DB types (Which I see how to do, sort of), our existing syntax defines DB types, that we should then map to Python types.

The big issue is they don't map cleanly. `str` is unbounded so technically `TextField`, but in reality it is more often a `CharField`.

We need a place to put wild ideas. An issue just doesn't feel right, but I don't see where else?

**Comment by abondar** [comment]
I think issues are fine for it, but we should formulate it so, that wild ideas should come with description of a problem that we are trying to address with this wild idea. So that way, discussion on how to overcome that problem would be in place and alternative solution for that problem could be found out in issue.

Regarding typing, at the moment I don't see how we can change our types for models because of static nature of `typing` module of python, while metaclasses work dynamically.
Django overcomes it with custom plugins for editors, but for Tortoise at the moment that seems like drastic overkill.

**Comment by grigi** [comment]
Yeah, I'm all for standardised solutions. e.g. when the `typing` plugin system gets pinned down we can have a try. This would be especially useful as many editors have standardised on that.

We already have a `PyLint` plugin that does some mapping, but that was to make `PyLint` useful for Tortoise.

I wonder if one can fool the typing system so that the type of `CharField(100, nullable=True)` returns a type of `Optional[str]`
In py3.6+ you can always define the python types yourself:
```py
name: Optional[str] = CharField(100, nullable=True)
```

**Comment by monospacedmagic** [comment]
Sounds good as long as we can still have attributes that aren't fields :) I have to use 3.6+ anyway because kennethreitz/responder is 3.6+ so I personally wouldn't be affected by a version bump. Just make sure there's still a way to manually define what type of Field an attribute is, i.e. an `int` could be a SmallIntField, an IntField or a BigIntField, for instance. (Not entirely sure if I understood this idea correctly. ^^' )

**Comment by antonagestam** [comment]
mypy seems to have a quite extensive plugin system, so something like this could probably me made to work. However, I think it would be a bad idea to give the fields types that aren't Python types. `str[255]` for instance doesn't make any sense, so projects using this ORM would have to ditch the builtin types as `str` would need a new definition.

If it's possible, I think it would be a better idea to tell the type system that the fields are subtypes of their respective Python types?

**Comment by grigi** [comment]
Yes, all valid points. This is feeling more like something that we may not want to do as is in my proposal.

Technically the problem is as follows:
Lets say I define a class `Stuff` with`val=IntField()`
It appears that for `Stuff` val is `Intfield` and a class variable.
But due to the metaclass rewriting the class, `Stuff` val is `int` and an instance variable.

In the name of being succinct (and use metaclasses as a solution) we will end up confusing anything that tries to do Type determinism.

I added a PyLint plugin so that it can stop complaining about everything, but at that time Mypy didn't have plugins yet.

I'd love a Mypy plugin :grin:

**Comment by antonagestam** [comment]
Cool! I think mypy support really would be a great feature.

The plugins for attrs and dataclasses that are shipped with mypy are probably good places to look for inspiration, and they should probably give a somewhat good indication of the amount of work that's required.

- dataclasses plugin: https://github.com/python/mypy/blob/00ccb783de8b43bcfcc22184da8846a927bcb9db/mypy/plugins/dataclasses.py
- attrs plugin: https://github.com/python/mypy/blob/00ccb783de8b43bcfcc22184da8846a927bcb9db/mypy/plugins/attrs.py
